src/tpm2_akparse.c

```diff
@@ -137,38 +137,36 @@ int SaveEccKeyToFile(const char keyfile[], UINT16 algType, const TPMS_ECC_POINT
 int parseAKPublicArea()
 {
     TPM2B_PUBLIC outPublic;
+    TPM2B *sizedBuffer;
     UINT16 size = sizeof(outPublic);
-    if( loadDataFromFile(akDataFile, (UINT8 *)&outPublic, &size) )
+
+    if ( loadDataFromFile(akDataFile, (UINT8 *)&outPublic, &size) )
     {
         return -1;
     }
 
-    if( TPM_ALG_RSA == outPublic.t.publicArea.type )
-    {
-        if(SaveKeyToFile(akKeyFile, TPM_ALG_RSA, outPublic.t.publicArea.unique.rsa.t))
-            return -2;
-    }
-    else if( TPM_ALG_ECC == outPublic.t.publicArea.type )
-    {
-        if(SaveEccKeyToFile(akKeyFile, TPM_ALG_ECC, outPublic.t.publicArea.unique.ecc))
-            return -3;
-    }
-    else if( TPM_ALG_KEYEDHASH == outPublic.t.publicArea.type )
-    {
-        if(SaveKeyToFile(akKeyFile, TPM_ALG_KEYEDHASH, outPublic.t.publicArea.unique.keyedHash.t))
-            return -4;
-    }
-    else if( TPM_ALG_SYMCIPHER == outPublic.t.publicArea.type )
+    switch ( outPublic.t.publicArea.type )
     {
-        if(SaveKeyToFile(akKeyFile, TPM_ALG_SYMCIPHER, outPublic.t.publicArea.unique.sym.t))
-            return -5;
-    }
-    else
-    {
-        printf("\nThe algorithm type(0x%4.4x) is not supported\n", outPublic.t.publicArea.type);
-        return -6;
+        case TPM_ALG_RSA:
+            sizedBuffer = (TPM2B *)&outPublic.t.publicArea.unique.rsa.t;
+        break;
+        case TPM_ALG_ECC:
+            sizedBuffer = (TPM2B *)&outPublic.t.publicArea.unique.ecc;
+        break;
```

Comment:
ECC case should be handled specially since ecc is not a TPM2B compatible structure. we should use SaveEccKeyToFile() to save it instead of SaveKeyToFile().